lib/hooks.ml

```diff
@@ -0,0 +1,18 @@
+let use_ref initial =
+  let state, _ = Core.use_state (fun () -> ref initial) in
+  state
+
+let use_effect ~on ?(release = Fun.const ()) acquire =
+  let last_value = use_ref on in
+  let state = use_ref None in
+  let release () = Option.iter release !state in
+  let acquire () = state := Some (acquire ()) in
+  Core.use_effect_once (fun () ->
+      acquire ();
+      Some release);
+  Core.use_effect_always (fun () ->
+      if on <> !last_value then (
```

Comment:
Can you think of a use case where you would actually want a shallow comparison where the shallow and deep comparison is not exactly the same? I can't think of any. And if there isn't any I can't see the value in making a distinction.

But you bring up a great point, and I think at the very least it would be a good idea to add an extra parameter to pass in an `equal` function, so that it's possible to customize the comparison for performance reasons.

Reply:
Added an `equal` parameter in https://github.com/ml-in-barcelona/jsoo-react/pull/154/commits/5961c6945e7550eb97607bda9b0e59cad5e4b816